### DISCUSSION TITLE
Property getters/setters and auto completion

### DISCUSSION THREAD

**Discussion by RedMser** [godotengine/godot-proposals#5220] [Scripting]
In C++, Godot uses getter and setter methods instead of properties.
In GDScript, you are encouraged to use the properties syntax only, but the getter/setter methods are also valid (even though they don't show auto completion anymore, see https://github.com/godotengine/godot/pull/56268).

This can leave a user (e.g. myself) confused, since some methods like `get_tree()` or `get_parent()` are *only* available as methods, while other properties have very similar sounding getters like `get_owner()` or `get_root()` and they do *not* show up in the auto completion of GDScript.

I understand and agree with removing the auto completion items for getter/setter methods, but this has tripped me up multiple times and I almost filed a bug report for it before realizing it's "intentional". Also confusing is that property getters/setters are [still included in the documentation pages](https://docs.godotengine.org/en/latest/classes/class_node.html#class-node-property-owner).

---

I have two potential solutions to help make this make more sense, but feel free to come up with others:

# 1) Remove the getter/setter table from the documentation

I mean this one:
![image](https://user-images.githubusercontent.com/5117197/185765062-cb98574a-3a4e-4224-8e8f-873d9f9cc4f4.png)

C++ users will find the corresponding methods by searching the auto completion for the property name, and GDScript users (the main target audience of these documentation pages) simply get confused by seeing these in here, since they don't seem to be valid.

Unless there's use cases for these method names (maybe using reflection to call the methods?), I don't think they belong in the docs anymore.

# 2) Make Godot's auto completion smarter

This idea would make it so that when typing the name of a getter/setter method, the auto completion for the corresponding property shows up. (Right now, no auto completion popup appears in this situation...)

Something like this:
![image](https://user-images.githubusercontent.com/5117197/185765174-a6356e7b-6f5a-4180-ba64-f3db0bcb5bc2.png)

No idea how this solution would work with LSP integration.

---

P.S. I don't know the situation in C#, but would be interested to hear if this is a problem there as well, or if all properties exist only as properties in C#.

**Comment by BaddRadish** [comment]
an exception could easily be made to remove the "get_" part when searching for a method/attribute, to make the search results work better in this way. im surprised they dont already show up.
